server/sonar-server/src/main/java/org/sonar/server/computation/component/PathAwareVisitorAdapter.java

```diff
@@ -0,0 +1,212 @@
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * SonarQube is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+package org.sonar.server.computation.component;
+
+import java.util.ArrayDeque;
+import java.util.Deque;
+import java.util.Iterator;
+import java.util.NoSuchElementException;
+
+import static java.util.Objects.requireNonNull;
+
+/**
+ * A adapter of the {@link PathAwareVisitor} to be able to visit only some component types
+ */
+public abstract class PathAwareVisitorAdapter<T> implements PathAwareVisitor<T>, Visitor {
+  private final Component.Type maxDepth;
+  private final Order order;
+  private final StackElementFactory<T> factory;
+  private final DequeBasedPathAdapter<T> stack = new DequeBasedPathAdapter<>();
+
+  public PathAwareVisitorAdapter(Component.Type maxDepth, Order order, StackElementFactory<T> factory) {
+    this.maxDepth = requireNonNull(maxDepth);
+    this.order = requireNonNull(order);
+    this.factory = requireNonNull(factory, "Factory can not be null");
+  }
+
+  public DequeBasedPath<T> getStack() {
+    return stack;
+  }
+
+  public void addComponentToStack(Component component){
+    stack.add(new PathAwareVisitorAdapter.PathElementImpl<>(component, createForComponent(component)));
+  }
+
+  public void popStack(){
+    stack.pop();
+  }
+
+  @Override
+  public Component.Type getMaxDepth() {
+    return maxDepth;
+  }
+
+  @Override
+  public Order getOrder() {
+    return order;
+  }
+
+  public StackElementFactory<T> getFactory() {
+    return factory;
+  }
+
+  protected T createForComponent(Component component) {
+    switch (component.getType()) {
+      case PROJECT:
+        return factory.createForProject(component);
+      case MODULE:
+        return factory.createForModule(component);
+      case DIRECTORY:
+        return factory.createForDirectory(component);
+      case FILE:
+        return factory.createForFile(component);
+      default:
+        return factory.createForUnknown(component);
+    }
+  }
+
+  public void visitProject(Component project, Path<T> path) {
+    // empty implementation, meant to be override at will by subclasses
+  }
+
+  public void visitModule(Component module, Path<T> path) {
+    // empty implementation, meant to be override at will by subclasses
+  }
+
+  public void visitDirectory(Component directory, Path<T> path) {
+    // empty implementation, meant to be override at will by subclasses
+  }
+
+  public void visitFile(Component file, Path<T> path) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Add the "@Override" annotation above this method signature [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1161)